updater/lib/dependabot/updater/operations/group_update_all_versions.rb

```diff
@@ -0,0 +1,370 @@
+#   effort to update everything it can in one pass.
+module Dependabot
+  class Updater
+    module Operations
+      class GroupUpdateAllVersions
+        GROUP_NAME_PLACEHOLDER = "*"
+
+        def self.applies_to?(job:)
+          return false if job.security_updates_only?
+          return false if job.updating_a_pull_request?
+          return false if job.dependencies&.any?
+
+          Dependabot::Experiments.enabled?(:grouped_updates_prototype)
+        end
+
+        def initialize(service:, job:, dependency_snapshot:, error_handler:)
+          @service = service
+          @job = job
+          @dependency_snapshot = dependency_snapshot
+          @error_handler = error_handler
+        end
+
+        def perform
+          Dependabot.logger.info("[Experimental] Starting grouped update job for #{job.source.repo}")
+          # We should log the rule being executed, let's just hard-code wildcard for now
+          # since the prototype makes best-effort to do everything in one pass.
+          Dependabot.logger.info("Starting update group for '#{GROUP_NAME_PLACEHOLDER}'")
+          dependency_change = compile_dependency_change
+
+          if dependency_change.dependencies.any?
+            Dependabot.logger.info("Creating a pull request for '#{GROUP_NAME_PLACEHOLDER}'")
+            begin
+              service.create_pull_request(dependency_change, dependency_snapshot.base_commit_sha)
+            rescue StandardError => e
+              # FIXME: This is a workround for not having a single Dependency to report against
+              #
+              #        We could use all_updated_deps.first, but that could be misleading. It may
+              #        make more sense to handle the group rule as a Dependancy-ish object
+              group_dependency = OpenStruct.new(name: "group-all")
+              raise if ErrorHandler::RUN_HALTING_ERRORS.keys.any? { |err| e.is_a?(err) }
+
+              error_handler.handle_dependabot_error(error: e, dependency: group_dependency)
+            end
+          else
+            Dependabot.logger.info("Nothing to update for Group Rule: '#{GROUP_NAME_PLACEHOLDER}'")
+          end
+        end
+
+        private
+
+        attr_reader :job,
+                    :service,
+                    :dependency_snapshot,
+                    :error_handler
+
+        def dependencies
+          all_deps = dependency_snapshot.dependencies
+
+          # Tell the backend about the current dependencies on the target branch
+          service.update_dependency_list(dependency_snapshot: dependency_snapshot)
+
+          allowed_deps = all_deps.select { |d| job.allowed_update?(d) }
+          # Return dependencies in a random order, with top-level dependencies
+          # considered first so that dependency runs which time out don't always hit
+          # the same dependencies
+          allowed_deps = allowed_deps.shuffle unless ENV["UPDATER_DETERMINISTIC"]
```

Comment:
I think we should probably remove shuffling _within_ groups and instead shuffle the groups when we have more than one, but for now I think shuffling has the benefit that it might flush out issues compiling changes together if we run them in a random order - if we made this deterministic now it could result in false confidence.